status.py

```diff
@@ -0,0 +1,62 @@
+from google.appengine.api import memcache
+
+class Status:
+    def __init__(self, fence_api, sam_api):
+        self.fence_api = fence_api
+        self.sam_api = sam_api
+
+    def get(self):
+        """
+        gets status of all sub systems
+        :return: dict with key the name of the subsystem, value is a dict with entries "ok": boolean and "message": string
+        """
+        try:
+            status = self._get_cached_status()
+            if status:
+                return status
+            else:
+                # if we got this far memcache is ok
+                fence_ok, fence_message = self.fence_api.status()
+                datastore_ok, datastore_message = self._datastore_status()
+                sam_ok, sam_message = self.sam_api.status()
+
+                status = {
+                    Subsystems.memcache: {"ok": True, "message": ""},
+                    Subsystems.datastore: {"ok": datastore_ok, "message": datastore_message},
+                    Subsystems.fence: {"ok": fence_ok, "message": fence_message},
+                    Subsystems.sam: {"ok": sam_ok, "message": sam_message}
+                }
+                self._cache_status(status)
+            return status
+        except Exception as e:
+            # any exception at this point is memcache
+            return {
+                Subsystems.memcache: {"ok": False, "message": e.message},
+                Subsystems.datastore: {"ok": False, "message": "unknown"},
```

Comment:
memcache being non-functional doesn't necessarily mean that the 3 other subsystems are down. we could still return the status for them. it'd be best to know the complete story during any outage

Reply:
I am going to leave it as is. If memcache is down and we query other systems we will hammer them.

Reply:
:+1: